Sources/Async/BackendService+Async.swift

```diff
@@ -0,0 +1,29 @@
+//
+//  BackendService+Async.swift
+//  Hyperspace-iOS
+//
+//  Created by Daniel Larsen on 10/30/21.
+//  Copyright © 2021 Bottle Rocket Studios. All rights reserved.
+//
+
+import Foundation
+
+extension BackendServiceProtocol {
+
+    /// Executes the Request, returning the specified value type. Throws an error if unable to return the value as specified.
+    ///
+    /// - Parameters:
+    ///   - request: The Request to be executed.
+    /// - Returns: The decodable type specified in the `Request`.
+    @available(iOSApplicationExtension 13.0.0, *)
+    public func execute<T, U>(request: Request<T, U>) async throws -> T {
```

Comment:
I would also like to see a non-throwing version of this that doesn't lose `U`'s type information. Something like 

```public func execute<T, U>(request: Request<T, U>) async -> Result<T, U>```

Doing it only this way, it feels like having the error type strongly typed on `Request` is useless, because you'll have to catch all errors anyway.

Reply:
Thanks so much for the feedback! I definitely agree that having strongly typed errors in the result is good, and being able to distinguish errors by type is useful.

Here's what I'm thinking, though. Doing a generic `do ... catch` could be a bit more true to the process. There are many different types of errors that could occur during the execution of the request in Hyperspace. Wouldn't it make sense to be able to throw all those errors as the type they are, not necessarily wrapped in another type?

But then we get to your question of why even specify the error type in the `Request`. And I agree that it becomes less useful, but it still might have a use when the response is an error code, but contains a valid body that can be decoded. 


To be completely honest, I feel like I'm missing something, though. Do you have any specific examples I can use to make sure that your use case is being covered?

Reply:
> Here's what I'm thinking, though. Doing a generic do ... catch could be a bit more true to the process. There are many different types of errors that could occur during the execution of the request in Hyperspace. Wouldn't it make sense to be able to throw all those errors as the type they are, not necessarily wrapped in another type?

I disagree- if I create a `Request<X, MyError>`, then I would expect to get back either `X` or an instance of `MyError`. Conformance of `MyError` to `TransportFailureRepresentable` is needed to ensure any transport failure can be converted to `MyError`, and optionally, conformance to `DecodingFailureRepresentable` is used to automatically handle JSON decoding errors.

In my opinion, the entire benefit of those two protocols is to encapsulate and surface the multitude of things that can go wrong in a single, easy-to-reason-with, type. If we think it would be better to throw all those errors as the type they are, then we could just extend `URLSession` to automatically try to decode something.

I'm not saying that something like `public func execute<T, U>(request: Request<T, U>) async -> Result<T, U>`, should be the only option (I think the majority of people who use `AnyError` would generally default to using a general `async throws` as they aren't getting much specific use out of `AnyError` anyway), but I do think it should be an option.

For example, let's say I have a type `MyError`. It conforms to `TransportFailureRepresentable`, so any time I execute a closure-based request, the compiler is guaranteeing the type of any error I receive. 

```swift
switch result {
case .success: break
case .failure(let error): //error is MyError
```

Using the `async` version though (basically because `throws` is currently untyped) I lose that guarantee, and have to handle a case that shouldn't be possible based on how `BackendService` is implemented.

```swift
let request<X, MyError> = ...
do {
 let myX = try await backendService.execute(request)
catch let error as MyError {
  // Handle my error
} catch {
  // Handle a case that shouldn't be possible, but I have to handle each time anyway
}
```

Reply:
Thanks so much for your thorough explanation, @wmcginty ! Given the use case you're laying out I do agree that a second function returning `Result<T, U>` does make the most sense. I'll add that and add unit tests for it as well.

I do think it'd be interesting to catch up on strategic approaches sometime. I still lean towards using the native `do ... catch` infrastructure over switching on a result, but I'd love to hear your experiences!

In the immediate term, though, I will add the async result function.

Reply:
Thanks@GrandLarseny! I see that approach as well, and I think when Hyperspace is fully built on `async` functions, the entire library might look a bit different, more geared to `do`, `try`, `catch`.